### ISSUE TITLE
starting a snapshot listener from afar causes problems 

### ISSUE THREAD

**Issue by michaelAtCoalesce** [firebase/firebase-js-sdk#8451]
### Operating System

windows

### Environment (if applicable)

chrome

### Firebase SDK Version

10.13.0

### Firebase SDK Product(s)

Firestore

### Project Tooling

create-react-app example 

### Detailed Problem Description

i have a collection with ~50 megabytes of data across ~1500 documents. 

when i try to start a listener while in north america (connecting to US firestore), it takes only 10 seconds to start the listener, and it completes 100% of the time

<img width="1066" alt="image" src="https://github.com/user-attachments/assets/9ff0dfef-4122-4ebb-b10b-7d720850f3fb">

. when i turn on my india VPN (same machine, same code, only difference is routing through india), the listener never even completes. 

i immediately get these errors - 

<img width="1030" alt="image" src="https://github.com/user-attachments/assets/d700a9e1-0018-4f76-a640-bb4d2a91c8ab">

on some machines from APAC region connecting to US firestore, i also get really poor behavior, and it never actually succeeds. 

![image](https://github.com/user-attachments/assets/7848e768-8e7e-485d-abb3-17b5da2c04a8)

for what its worth - the connection FROM india to united states should easily be able to handle this...

<img width="747" alt="image" src="https://github.com/user-attachments/assets/ecce280f-cfc3-4f76-b4b7-e0e80a398756">

### Steps and code to reproduce issue

if someone wants to email me i can send them the info for the recreate. it's a few lines of code.

**Comment by michaelAtCoalesce** [comment]
on slower machines from APAC connecting to north america,  we will also get 400's. either way, looks like the listeners either take forever, or are erroring it 

<img width="1036" alt="image" src="https://github.com/user-attachments/assets/c421f285-bdb0-4d47-af16-1e7395bee59e">

**Comment by michaelAtCoalesce** [comment]
based on previous issues @dconeybe might be a good person to look at this? Someone can email me at 'mx2323 <@> gmail.com' and I'll  jump on a call to recreate.

**Comment by michaelAtCoalesce** [comment]
any updates? this is causing our application to not load.

**Comment by wu-hui** [comment]
Hey @michaelAtCoalesce ,

I suspect the issue here is that the bidirectional stream between the SDK and the backend does not work well when the network is not stable, especially when you need to load a lot data over the wire. There are several things you can try:

1. Can you create a test firestore in Asian to see if things improve?
2. Try to always turn on longpolling (https://firebase.google.com/docs/reference/js/firestore_.firestoresettings.md#firestoresettingsexperimentalforcelongpolling) see if that helps.

@sampajano Do you have some other suggestions/ideas?

**Comment by michaelAtCoalesce** [comment]
hi @wu-hui, 

1. yes, this is definitely the case. as i said above, when i turn off the VPN (so connect from nearby) - the reliability is 100% 
2. i have forced long polling, but i am noticing that i see slower load times. more consistent, but slower load times. 

the data here is only on the order of tens of megabytes - but it can take over a minute (and sometimes not load at all), whereas a closer connection will take ~10 seconds. 

in conclusion: 
the speed test shows that this should work, so it seems there is work to do here for firestore to reliability support this kind of a connection... it shouldn't take over a minute to load when it takes 10 seconds in the ideal case and even the force long polling is slow. 

i can reliably recreate this issue 100% of the time

**Comment by michaelAtCoalesce** [comment]
why was the needs attention label removed and a needs-info label added? i believe ive given the information required and this is causing our production to not load.

**Comment by DellaBitta** [comment]
My mistake! I must have had been looking at a stale page that I had loaded yesterday, sorry!

**Comment by michaelAtCoalesce** [comment]
from nearby: 11seconds
from afar: 226 seconds

im uploading some firestore debug level logs of the degenerative case here. 

[aec2-38-34-123-154.ngrok-free.app-1724779566040.log](https://github.com/user-attachments/files/16766539/aec2-38-34-123-154.ngrok-free.app-1724779566040.log)

**Comment by michaelAtCoalesce** [comment]
any updates?

**Comment by MarkDuckworth** [comment]
@michaelAtCoalesce, Thank you for providing the logs. I reviewed them and I don't see a clear indication of an issue in the SDK, however I have forwarded this to our backend team for review. Googlers see b/361143373

For what it's worth, on this behavior, you may get more frequent updates if you open a Firebase or Google Cloud support ticket rather than a GitHub issue on the SDK. However, we will update this GH issue when we learn more.

**Comment by michaelAtCoalesce** [comment]
@MarkDuckworth thanks for the update. wanted to add another data point in here. it appears to happen more frequently on windows. i have anecdotally noticed in my recreate case that on chrome on windows the default implementation is more likely to fail than chrome on macOS. if nothing else, it appears that windows is at least 3x as slow.

**Comment by michaelAtCoalesce** [comment]
its been 2 weeks.. any updates? 

i was told by firebase support to try to paginate the snapshots. i tried that, and the performance did not improve.

**Comment by vdemko001** [comment]
Any updates?
I'm getting the same error on Firebase SDK 7.24.0
![image](https://github.com/user-attachments/assets/559d165b-0b43-41bc-9cec-9e3d305b6b35)

**Comment by mx2323** [comment]
@MarkDuckworth I think the enabling of web channel in 10.14.0 is improving the performance of my standalone test case. Sounds like with large datasets and chunking there may have been a corruption causing memory issues as a symptom that was fixed. 

We are still having an issue where inexplicably firestore will take 50 seconds within our app to listen on windows, but in the standalone test case on windows it’ll take 15 seconds, consistently. Our users frequently open and close listeners as part of their workflow so the performance is important for them. 

once I get some more data will open a ticket for that

**Comment by MarkDuckworth** [comment]
@mx2323, thanks for the feedback on 10.14.0. It's good to hear it is helping you out.

Regarding the 50 second delay, is it consistently 50 seconds? Take a look at https://github.com/firebase/firebase-js-sdk/issues/8474, it could be related to that. Although we have only seen that delay at ~45 seconds. Your logs in https://github.com/firebase/firebase-js-sdk/issues/8451#issuecomment-2313138710 don't show similarity, but if you're seeing a consistent 45-50 second delay, it may be worth getting logs covering this timespan.

**Comment by mx2323** [comment]
ok I’ll follow up soon. headed on a flight and will be unavailable.

**Comment by michaelAtCoalesce** [comment]
standalone recreate - 11 seconds 
[fastmac-11seconds.log](https://github.com/user-attachments/files/17264048/fastmac-11seconds.log)

<img width="1025" alt="image" src="https://github.com/user-attachments/assets/65cebe56-b6ed-4fb7-87ce-47fe48fe7651">


standalone recreate on windows india vm - 40 seconds
[standalone-windows-40seconds.log](https://github.com/user-attachments/files/17264104/standalone-windows-40seconds.log)

in-app recreate on windows india vm - 75 seconds
[windows-insideapp-75seconds.log](https://github.com/user-attachments/files/17264106/windows-insideapp-75seconds.log)
<img width="1039" alt="image" src="https://github.com/user-attachments/assets/8f58abe0-f926-44f7-b894-1563fb1668cc">

**Comment by mx2323** [comment]
turning off autodetect long polling has made the performance worse across the board. log attached. 

seems like, if the autodetectlongpolling is turned on, and it doesnt detect a buffering proxy, thats the best performance. 
[120seconds-autodetectlongpollingoff.log](https://github.com/user-attachments/files/17265040/120seconds-autodetectlongpollingoff.log)

**Comment by michaelAtCoalesce** [comment]
Is there someone I can send a recreate to? What are next steps here?

**Comment by wu-hui** [comment]
One way to share the reproduction is to create a private repo, and invite me and @sampajano to join. 

Please provide a clear instruction on how to reproduce, especially this seems to involve some VM setup and VPNs. 

Also, we have no control over VPN or public internet, speed test does not necessarily translate to actual connection. We will do our best to look into this, but it is certainly possible that this won't lead us anywhere.

**Comment by michaelAtCoalesce** [comment]
@MarkDuckworth what we noticed and had filed a ticket for originally was that the client sdk was very unreliable in windows chrome browsers with slightly larger datasets.

We noticed that the backend Firestore client has a preferRest option that was much faster and more reliable.  We noticed that although the rest api was faster, we still had issues where concurrent requests within the same backend process were slow. I filed a ticket about this here https://github.com/googleapis/nodejs-firestore/issues/2215. Ultimately I think the unzipping of hundreds of megabytes of data from the rest api is just slow and hogs the main thread because forking processes and executing on the same machine was parallel and fast. 


After all this investigation, what we decided to do is reverse engineer with wireshark what the backend was doing with preferRest (since the runQuery docs are not helpful https://firebase.google.com/docs/firestore/reference/rest/v1beta1/projects.databases.documents/runQuery)  and call that via the frontend directly. Then we have changed every write in our app to store server timestamps, and we changed our snapshot listener to filter for all values higher than the values seen than what we pulled down in the starting snapshot. We also implemented a separate collection so we can have deletes. 

So far in our extensive testing the result has more reliable and consistent performance than the Firestore client web sdk. It's unfortunate that we had to do this, but I repeatedly reached out to Firestore team members over many months and sent logs and offered multiple times to jump on a call with no luck.

**Comment by wu-hui** [comment]
Thanks for the update, it is great that you found a way that works for you.

I have couple questions:

1. runQuery does not support realtime queries, how do you use snapshot listener still?
2. About "Then we have changed every write in our app to store server timestamps, and we changed our snapshot listener to filter for all values higher than the values seen than what we pulled down in the starting snapshot.". Does this mean eventually you get the improvements by having less documents sent to the client?
